### PR TITLE
KEP-3619: Fine-grained SupplementalGroups control

### PR DESCRIPTION
**Pull request by everpeace**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Implements https://github.com/kubernetes/enhancements/issues/3619

#### Special notes for your reviewer:

The actual policy enforcement will be implemented by container runtimes (cri-api implementations). Because most container runtimes (e.g. containerd) depend on https://github.com/kubernetes/cri-api, this PR should be merged before container runtimes implement actual policy enforcement. Please also see [CRI API | Feature development](https://github.com/kubernetes/kubernetes/blob/6546c8b/staging/src/k8s.io/cri-api/README.md#feature-development).

My PR in containerd is already opened and in review: https://github.com/containerd/containerd/pull/9737. It already got LGTM (https://github.com/containerd/containerd/pull/9737#pullrequestreview-2057591573) except for cri-api dependency, which will be updated once this PR gets merged and v1.31 alpha or beta version is released.

_If you wanted to try running this feature, you can try it with kind easily: https://gist.github.com/everpeace/2ae0233cc91644ac8797cf192e40ba39_

#### Does this PR introduce a user-facing change?

```release-note
Support fine-grained supplemental groups policy (KEP-3619), which enables fine-grained control for supplementary groups in the first container processes. You can choose whether to include groups defined in the container image(/etc/groups) for the container's primary uid or not.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3619-supplemental-groups-policy
```

### CODE REVIEW

**Comment by everpeace** [comment]
/cc  @thockin @mrunalp @SergeyKanzhelev

Let me add you guys as [this KEP reviewers](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/3619-supplemental-groups-policy/kep.yaml#L10C1-L15)

**Comment by everpeace** [comment]
rebased.

File: pkg/apis/autoscaling/validation/validation.go [hunk 1]
```diff
@@ -18,6 +18,7 @@ package validation
 
 import (
 	"fmt"
+
```
Comment (thockin) [inline pkg/apis/autoscaling/validation/validation.go:21]: can you just drop this file?

File: pkg/apis/core/types.go [hunk 2]
```diff
@@ -2505,6 +2505,32 @@ type ContainerStatus struct {
...
+	// +featureGate=SupplementalGroupsPolicy
+	// +optional
+	User *ContainerUser
+}
+
+// ContainerUser represents user identity information
+type ContainerUser struct {
+	// Linux holds user identity information of the first process of the containers in Linux.
+	// Note that this field cannot be set when spec.os.name is windows.
+	Linux *LinuxContainerUser
```
Comment (thockin) [inline pkg/apis/core/types.go:2519]: I don't love spelling out Linux here - is there precedent?

File: pkg/apis/core/validation/validation.go [hunk 3]
```diff
@@ -4421,6 +4424,24 @@ func validateSysctls(sysctls []core.Sysctl, fldPath *field.Path) field.ErrorList
 	return allErrs
 }
 
+var supportedSupplementalGroupsPolicy = sets.NewString(string(core.SupplementalGroupsPolicyMerge), string(core.SupplementalGroupsPolicyStrict))
+
+func validateSupplementalGroupsPolicy(supplementalGroupsPolicy *core.SupplementalGroupsPolicy, podOs *core.PodOS, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.SupplementalGroupsPolicy) {
```
Comment (thockin) [inline pkg/apis/core/validation/validation.go:4432]: This is not how we usually gate new fields.

The gate controls admission into the system.  Once it is in, we use it.  What that means practically is that you want to add this in k8s.io/kubernetes/pkg/api/pod `DropDisabledPodFields()` (see the other examples).

Then, in here, you do something like:

```
if gate enabled {
    if field is nil {
        field.Required(...)
    } else if field is not a valid value {
            field.Invalid(...)
    }
}
```

You also need to make sure that no updates are allowed when the gate is disabled.

File: pkg/apis/core/validation/validation.go [hunk 4]
```diff
@@ -4775,6 +4798,15 @@ func ValidatePodUpdate(newPod, oldPod *core.Pod, opts PodValidationOptions) fiel
 // ValidateContainerStateTransition test to if any illegal container state transitions are being attempted
 func ValidateContainerStateTransition(newStatuses, oldStatuses []core.ContainerStatus, fldpath *field.Path, restartPolicy core.RestartPolicy) field.ErrorList {
 	allErrs := field.ErrorList{}
+
+	if !utilfeature.DefaultFeatureGate.Enabled(features.SupplementalGroupsPolicy) {
```
Comment (thockin) [inline pkg/apis/core/validation/validation.go:4802]: same as above, except this is not required

Comment (thockin) [reply pkg/apis/core/validation/validation.go:4802]: Actually, I am not sure there's any validation you CAN do here.  Maybe check thart UID and GIDs are not negative?

File: pkg/apis/core/types.go [hunk 5]
```diff
@@ -2505,6 +2505,32 @@ type ContainerStatus struct {
...
+
+	// Windows holds user identity information of the first process of the containers in Windows
+	// This is just reserved for future use.
+	// Windows *WindowsContainerUser
+}
+
+// LinuxContainerUser represents user identity information in Linux containers
+type LinuxContainerUser struct {
+	// Uid is the primary uid of the first process in the container
+	Uid int64
```
Comment (thockin) [inline pkg/apis/core/types.go:2529]: UID and GID

File: staging/src/k8s.io/api/core/v1/types.go [hunk 6]
```diff
@@ -2723,6 +2723,34 @@ type ContainerStatus struct {
...
+
+	// Windows holds user identity information of the first process of the containers in Windows
+	// This is just reserved for future use.
+	// Windows *WindowsContainerUser
+}
+
+// LinuxContainerUser represents user identity information in Linux containers
+type LinuxContainerUser struct {
+	// Uid is the primary uid of the first process in the container
+	Uid int64 `json:"uid,omitempty" protobuf:"varint,1,name=uid"`
```
Comment (thockin) [inline staging/src/k8s.io/api/core/v1/types.go:2747]: don't omitempty - we want 0 to print

File: staging/src/k8s.io/api/core/v1/types.go [hunk 7]
```diff
@@ -2723,6 +2723,34 @@ type ContainerStatus struct {
...
+	// Gid is the primary gid of the first process in the container
+	Gid int64 `json:"gid,omitempty" protobuf:"varint,2,name=gid"`
```
Comment (thockin) [inline staging/src/k8s.io/api/core/v1/types.go:2749]: same

File: staging/src/k8s.io/api/core/v1/types.go [hunk 8]
```diff
@@ -3788,9 +3836,17 @@ type PodSecurityContext struct {
...
+	// If SupplementalGroupsPolicy feature gate was enabled, you can control how the groups are calculated.
 	// Note that this field cannot be set when spec.os.name is windows.
 	// +optional
 	SupplementalGroups []int64 `json:"supplementalGroups,omitempty" protobuf:"varint,4,rep,name=supplementalGroups"`
+	// Defines how supplemental groups of the first container processes are calculated.
+	// Valid values are "Merge" and "Strict". If not specified, "Merge" is used.
+	// (Alpha) Using the field requires the SupplementalGroupsPolicy feature gate to be enabled
+	// and the container runtime must implement support for this feature.
+	// +featureGate=SupplementalGroupsPolicy
+	// +optional
```
Comment (thockin) [inline staging/src/k8s.io/api/core/v1/types.go:3848]: I think you can `+default=Merge` here and avoid manual defaulting

**Review by thockin — COMMENTED** [review]
You need to set the default value in pkg/apis/core/v1/defaults.go.

**Comment by thockin** [comment]
I am reviewing PRs today and Monday, and then I am out - if we want to get this into 1.28, it needs to be updated ASAP.

**Comment by thockin** [comment]
I see this got missed - is it still something we want?  I think it's a good change, we just need to revisit the PR....

**Comment by everpeace** [comment]
> I think it's a good change, we just need to revisit the PR....

Thanks, I agree.

---

Sorry for my long absence. Let me revisit this PR in the next week.

This PR includes:
- CRI changes
- k8s API changes
- Implementation
  - k8s api validation
  - wiring SupplementalGroupPolicy/LinuxUser to CRI
  - bubbling up LinuxContainerUser from CRI to k8s api

However, no container runtime does support this feature currently. So, I'm thinking I should split up this PR into several PRs

- PR 1: CRI change
- PR 2: implementation the feature in containerd or CRI-O 
- PR 3: k8s API changes and implementation

What do you think??

**Comment by thockin** [comment]
The problem with breaking a PR up like that is that it is SO EASY to get stuck in the middle with something that can't be used but also can't easily be reverted.  I'd rather it go in all at once, even if there's no runtime support yet, and stay alpha until "enough" runtimes handle it

**Comment by thockin** [comment]
Also, we might hold off on merging it until at least one major runtime supports it, so we can ACTUALLY prove it.

**Comment by everpeace** [comment]
I see, I understood your point. And, I also found this document: https://github.com/kubernetes/cri-api?tab=readme-ov-file#feature-development

This document says:
> The recommended feature development flow is following::
> - Review proposed CRI API changes during the KEP review stage. Some field names and types may not be spelled out exactly at this stage.
> - Locally implement a prototype that implement changes in both - Kubernetes and Container Runtime.
> - Submit a Pull Request for Kubernetes implementing CRI API changes alongside the feature code. Feature must be developed to degrade gracefully when used with older Container Runtime according to the Version Skew policy.

Let me follow this path. I will keep working on this PR.

File: pkg/apis/core/types.go (hunk 2 above)
Comment (everpeace) [reply pkg/apis/core/types.go:2519]: @thockin sorry for long delay.

> is there precedent?

As far as I know, probably `seLinuxOptions` in `SecurityContext`?

However, I'm not a Windows specialist, though, UID/GID is a specific concept in Linux, isn't it? Then, I don't have a good idea to express this information in the container status.

Would you have suggestions?

**Comment by thockin** [comment]
Pinging this old one, which I still would like to see happen.

**Comment by everpeace** [comment]
Sorry for my absence the long time. Let me resume my work. I will update my branch soon.

**Comment by everpeace** [comment]
I merged master branch for ease of review. I will rebase to the latest master once review is approved.

If it is better to rebase, I would be happy if you tell me so 🙇

File: staging/src/k8s.io/api/core/v1/types.go (hunk 8 above)
Comment (everpeace) [reply staging/src/k8s.io/api/core/v1/types.go:4189]: I think `SupplementalGroupsPolicy` is os dependent. We should reject this field when os is windows. So, I didn't put the `+default` tag and keeps windows validation.

https://github.com/kubernetes/kubernetes/pull/117842/files#diff-c713e8919642d873fdf48fe8fb6d43e5cb2f53fd601066ff53580ea655948f0dR4283-R4285

File: pkg/apis/core/types.go (hunk 2 above)
Comment (everpeace) [reply pkg/apis/core/types.go:2712]: @thockin I keep this as is because I still have no idea how to avoid spelling out Linux here because supplementalGroups is the idea which is specific to Linux🙇 Any suggestions??

**Review by everpeace — COMMENTED** [review]
~~@thockin @mrunalp @SergeyKanzhelev Sorry for my long absence.~~

~~I updated my branch(just merged master for smooth review) and dropped my old validation logic and implemented `DropDisabledPodFields` as @thockin [pointed out](https://github.com/kubernetes/kubernetes/pull/117842/files/2808dde105103b08498a687b5d3fbdaf67f6dff8#r1210928751). I'm also working on cri implementation in https://github.com/containerd/containerd/pull/9737~~

~~You can try this feature with kind: https://gist.github.com/everpeace/2ae0233cc91644ac8797cf192e40ba39#file-readme-md~~

~~Please take a look 🙇~~

~~Note:~~
~~- Once got final approval, I will rebase to the latest master~~

**Review by everpeace — COMMENTED** [review]
@thockin @mrunalp @SergeyKanzhelev (KEP reviewers)

Hi, let me reboot this PR. I would like to aim for this feature to be released in v1.31. 

I rebased my branch to the latest master because my branch was too stale to merge back to it. Could you review this PR again🙇🙇? 

As you know, this KEP requires implementing CRI on the container runtime side. I already opened a PR https://github.com/containerd/containerd/pull/9737 as the first supported runtime for alpha release [as described in KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3619-supplemental-groups-policy#alpha). 

_Added: the https://github.com/containerd/containerd/pull/9737 got LGTM(https://github.com/containerd/containerd/pull/9737#pullrequestreview-2057591573) except for cri-api dependency._

_If you wanted to try running this feature, you can try it with kind easily: https://gist.github.com/everpeace/2ae0233cc91644ac8797cf192e40ba39_

File: pkg/api/pod/util.go [hunk 9]
```diff
@@ -809,16 +818,16 @@ func dropDisabledPodStatusFields(podStatus, oldPodStatus *api.PodStatus, podSpec
...
 
-	if !utilfeature.DefaultFeatureGate.Enabled(features.RecursiveReadOnlyMounts) && !rroInUse(oldPodSpec) {
-		for i := range podStatus.ContainerStatuses {
-			podStatus.ContainerStatuses[i].VolumeMounts = nil
-		}
-		for i := range podStatus.InitContainerStatuses {
-			podStatus.InitContainerStatuses[i].VolumeMounts = nil
-		}
-		for i := range podStatus.EphemeralContainerStatuses {
-			podStatus.EphemeralContainerStatuses[i].VolumeMounts = nil
```
Comment (ktock) [inline pkg/api/pod/util.go:820]: Is this change related to this PR?

Comment (everpeace) [reply pkg/api/pod/util.go:820]: Thanks for the catch! I put this code back🙇

Comment (everpeace) [reply pkg/api/pod/util.go]: addressed in 2db0c01. sorry for the rebase 🙇

File: pkg/apis/core/types.go [hunk 10]
```diff
@@ -3629,9 +3673,17 @@ type PodSecurityContext struct {
 	// no additional groups are added to any container. Note that group memberships
 	// defined in the container image for the uid of the container process are still effective,
 	// even if they are not included in this list.
+	// If SupplementalGroupsPolicy feature was enabled, you can control how the groups are calculated.
 	// Note that this field cannot be set when spec.os.name is windows.
 	// +optional
 	SupplementalGroups []int64
+	// Defines how supplemental groups of the first container processes are calculated.
+	// Valid values are "Merge" and "Strict". If not specified, "Merge" is used.
```
Comment (thockin) [inline pkg/apis/core/types.go:3681]: Previously we said we cannot have a literal default value because of bugs, but once https://github.com/kubernetes/kubernetes/pull/124220 lands, we can actually do this properly.

**Comment by thockin** [comment]
I am still keen to see this go in!  Thanks for keeping it going @everpeace !

File: pkg/apis/core/types.go [hunk 11]
```diff
@@ -2741,6 +2741,32 @@ type ContainerStatus struct {
...
+	// +optional
+	User *ContainerUser
+}
+
+// ContainerUser represents user identity information
+type ContainerUser struct {
+	// Linux holds user identity information initially attached to the first process of the containers in Linux.
+	// Note that the actual process identity can be dynamic if the initially attached identity have enough privilege calling setuid/setgid/setgroups syscalls
+	// +optional
+	Linux *LinuxContainerUser
```
Comment (everpeace) [inline pkg/apis/core/types.go:2755]: I emphasize the information is __initially attached__ one here.

As @ktock pointed out in https://github.com/containerd/containerd/pull/9737#discussion_r1598603468, process identity can be dynamic/updated if the initially attached identity has enough privilege calling setuid/setgid/setgroups syscall.

File: pkg/apis/core/types.go (hunk 10 above)
Comment (thockin) [reply pkg/apis/core/types.go:3682]: The bug today is that a new field with a default value gets considered by kubelet as a change, so it restarts all pods (and all nodes do this roughly in sync).  #124220  fixes that, so a new default should not trigger a restart.

> this field should be an OS dependent field

OK, thast may still be true, but adding a default will avoid code like:

```
if windows {
  return doSomethingElse()
}
policy = merge // hardcoded default
if FSGroupChangePolicy == nil {
    policy = *FSGroupChangePolicy
}
return doSomethingWithPolicy(policy)
```

And when users `kubectl get` the pods they will see "Merge" instead of nothing

Comment (everpeace) [reply pkg/apis/core/types.go:3682]: > so a new default should not trigger a restart

Wow, Thanks. That is why we have such constraint in `default_test.go`, right?

https://github.com/kubernetes/kubernetes/blob/5f1e5cafd33a8863c5dfe9f36c9cba836f0e981b/pkg/apis/core/v1/defaults_test.go#L165-L174

I didn’t know that. I’ve learned something today.

<hr />

> but adding a default will avoid code like:
> And when users kubectl get the pods they will see "Merge" instead of nothing

Thanks for this, too. Probably, now I got your idea. Once #124220 got landed, we can introduce to add new field with default value. But, this time, we would like to introduce conditional(os specific) default value like this:

- when `spec.os.name == windows`, it should validate the field must be `nil` 
- when `spec.os.name == linux`, it should make the default value be `Merge`

Then, perhaps, we can use custom defaulter function instead of `+default=“Merge”` annotation to achieve conditional default value?? I imagine like below:

```golang
# in pkg/apis/core/v1/defaults.go
func SetDefaults_PodSpec(obj *v1.PodSpec) {
…

	if obj.SecurityContext == nil {
		obj.SecurityContext = &v1.PodSecurityContext{}
	}
	if utilfeature.DefaultFeatureGate.Enabled(features.SupplementalGroupsPolicy) {
		if obj.SecurityContext.SupplementalGroupsPolicy == nil && (obj.OS == nil || obj.OS.Name == v1.Linux) {
			merge := v1.SupplementalGroupsPolicyMerge
			obj.SecurityContext.SupplementalGroupsPolicy = &merge
		}
	}
…
```

Comment (thockin) [reply pkg/apis/core/types.go:3682]: @alexzielenski Here's a fun case for declarative defaults  :)

@everpeace - something like that, yes.

**Comment by thockin** [comment]
This needs sig-node review and approval:  @SergeyKanzhelev @mrunalp @dchen1107 ?

File: pkg/apis/core/types.go [hunk 12]
```diff
@@ -2741,6 +2741,32 @@ type ContainerStatus struct {
...
+	// User represents user identitiy information initially attached to the first process of the container
+	// +featureGate=SupplementalGroupsPolicy
+	// +optional
+	User *ContainerUser
+}
+
+// ContainerUser represents user identity information
+type ContainerUser struct {
+	// Linux holds user identity information initially attached to the first process of the containers in Linux.
+	// Note that the actual process identity can be dynamic if the initially attached identity have enough privilege calling setuid/setgid/setgroups syscalls
```
Comment (thockin) [inline pkg/apis/core/types.go:2753]: Propose a reword:

```
Note that the actual running identity can be changed if the process has enough privilege to do so.
```

File: pkg/apis/core/types.go (hunk 2 above)
Comment (thockin) [reply pkg/apis/core/types.go:2755]: @liggitt @tallclair for consult.  I hate that the API spells out "Linux" but the structure would be different for Windows.

File: pkg/apis/core/types.go [hunk 13]
```diff
@@ -3541,6 +3568,24 @@ const (
 	FSGroupChangeAlways PodFSGroupChangePolicy = "Always"
 )
 
+// SupplementalGroupsPolicy defines how supplemental groups
+// of the first container processes are calculated.
+type SupplementalGroupsPolicy string
+
+const (
+	// Merge policy always merges the provided SupplementalGroups (including FsGroup)
```
Comment (thockin) [inline pkg/apis/core/types.go:3576]: Wording:

```
	// SupplementalGroupsPolicyMerge means that the container's provided
	// SupplementalGroups and FsGroup (specified in SecurityContext) will be
	// merged with the primary user's groups as defined in the container image
	// (in /etc/group).
```

File: pkg/apis/core/types.go [hunk 14]
```diff
@@ -3541,6 +3568,24 @@ const (
...
+	// specified in SecurityContext with groups of the primary user from the container
+	// image(`/etc/group`).
+	// Note: The primary user is specified with RunAsUser.
```
Comment (thockin) [inline pkg/apis/core/types.go:3579]: I think we can drop this part, or put it on the RunAsUser field

File: pkg/apis/core/types.go [hunk 15]
```diff
@@ -3541,6 +3568,24 @@ const (
...
+	//       If not specified, the user from the image config is used.
+	//       Otherwise, the runtime default is used.
+	SupplementalGroupsPolicyMerge SupplementalGroupsPolicy = "Merge"
+	// Strict policy uses only the provided SupplementalGroups(including FsGroup)
```
Comment (thockin) [inline pkg/apis/core/types.go:3583]: Wording

```
	// SupplementalGroupsPolicyStrict means that the container's provided
	// SupplementalGroups and FsGroup (specified in SecurityContext) will be
	// used instead of any groups defined in the container image.
```

File: pkg/apis/core/types.go [hunk 16]
```diff
@@ -3629,9 +3674,18 @@ type PodSecurityContext struct {
 	// no additional groups are added to any container. Note that group memberships
 	// defined in the container image for the uid of the container process are still effective,
 	// even if they are not included in this list.
+	// If SupplementalGroupsPolicy feature was enabled, you can control how the groups are calculated.
```
Comment (thockin) [inline pkg/apis/core/types.go:3677]: Can we reword this wole comment?

```
	// A list of groups applied to the first process run in each container, in
	// addition to the container's primary GID and fsGroup (if specified).  If
	// the SupplementalGroupsPolicy feature is enabled, the
	// supplementalGroupsPolicy field determines whether these are in addition
	// to or instead of any group memberships defined in the container image.
	// If unspecified, no additional groups are added, though group memberships
	// defined in the container image may still be used, depending on the
	// supplementalGroupsPolicy field.
```

**Review by everpeace — COMMENTED** [review]
Thank you for re-wording suggestions. Addressed in below packages:
- `pkg/core/v1`
- `staging/k8s.io/api/core/v1`
- `staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto`

File: pkg/apis/core/types.go (hunk 10 above)
Comment (liggitt) [reply pkg/apis/core/types.go:3681]: > once https://github.com/kubernetes/kubernetes/pull/124220 lands, we can actually do this properly

well... once it lands and makes it to the oldest kubelet that could be running against the server (so once the n-3 kubelet is 1.31, so 1.34)

File: pkg/apis/core/types.go (hunk 2 above)
Comment (liggitt) [reply pkg/apis/core/types.go:2755]: It is ugly, but I agree the shape for linux and windows is usually different.

For prior art, we have `securityContext.windowsOptions` at the pod and container level. In the cri-api, we also have 

```
    // Optional configurations specific to Linux hosts.
    LinuxPodSandboxConfig linux = 8;
    // Optional configurations specific to Windows hosts.
    WindowsPodSandboxConfig windows = 9;
```

For OS-specific fields like this, do validate that this is not set if `spec.podOS.name` is set to something other than "linux"

File: pkg/apis/core/types.go (hunk 10 above)
Comment (thockin) [reply pkg/apis/core/types.go:3681]: oh, yeah, I guess that's a good point.  I got a little ahead of myself.  Sorry @everpeace - what we CAN do is leave "TODO" comments to revisit this in 1.34 :)

File: pkg/apis/core/types.go (hunk 2 above)
Comment (thockin) [reply pkg/apis/core/types.go:2755]: ACK

File: pkg/apis/core/types.go [hunk 17]
```diff
@@ -2741,6 +2741,32 @@ type ContainerStatus struct {
...
+	// +optional
+	User *ContainerUser
+}
+
+// ContainerUser represents user identity information
+type ContainerUser struct {
+	// Linux holds user identity information initially attached to the first process of the containers in Linux.
+	// Note that the actual running identity can be changed if the process has enough privilege to do so.
+	// +optional
+	Linux *LinuxContainerUser
```
Comment (thockin) [inline pkg/apis/core/types.go:2755]: I think we need validation logic hat this is not set when the OS is Windows

File: pkg/apis/core/types.go [hunk 18]
```diff
@@ -3623,15 +3666,26 @@ type PodSecurityContext struct {
...
 	SupplementalGroups []int64
+	// Defines how supplemental groups of the first container processes are calculated.
+	// Valid values are "Merge" and "Strict". If not specified, "Merge" is used.
+	// (Alpha) Using the field requires the SupplementalGroupsPolicy feature gate to be enabled
+	// and the container runtime must implement support for this feature.
+	// Note that this field cannot be set when spec.os.name is windows.
+	// TODO: update the default value to "Merge" when spec.os.name is not windows in v1.34
+	// +featureGate=SupplementalGroupsPolicy
+	// +optional
+	SupplementalGroupsPolicy *SupplementalGroupsPolicy
```
Comment (thockin) [inline pkg/apis/core/types.go:3688]: I don't see validation code for this?

**Review by thockin — COMMENTED** [review]
API LGTM, just a couple small detals.

**Review by everpeace — COMMENTED** [review]
Sorry for the lack of validation logic. Addressed. PTAL 🙇

File: pkg/apis/core/validation/validation.go [hunk 19]
```diff
@@ -4318,6 +4318,11 @@ func validateWindows(spec *core.PodSpec, fldPath *field.Path) field.ErrorList {
 		if securityContext.SupplementalGroups != nil {
 			allErrs = append(allErrs, field.Forbidden(fldPath.Child("securityContext").Child("supplementalGroups"), "cannot be set for a windows pod"))
 		}
+		if utilfeature.DefaultFeatureGate.Enabled(features.SupplementalGroupsPolicy) {
```
Comment (thockin) [inline pkg/apis/core/validation/validation.go:4321]: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/feature-gates.md#using-a-feature-gate-in-code

Dont need to check gates here.

File: pkg/apis/core/validation/validation.go [hunk 20]
```diff
@@ -4948,6 +4953,12 @@ func validatePodSpecSecurityContext(securityContext *core.PodSecurityContext, sp
 		allErrs = append(allErrs, validateSeccompProfileField(securityContext.SeccompProfile, fldPath.Child("seccompProfile"))...)
 		allErrs = append(allErrs, validateWindowsSecurityContextOptions(securityContext.WindowsOptions, fldPath.Child("windowsOptions"))...)
 		allErrs = append(allErrs, ValidateAppArmorProfileField(securityContext.AppArmorProfile, fldPath.Child("appArmorProfile"))...)
+
+		if utilfeature.DefaultFeatureGate.Enabled(features.SupplementalGroupsPolicy) {
```
Comment (thockin) [inline pkg/apis/core/validation/validation.go:4957]: Same

File: pkg/apis/core/validation/validation.go [hunk 21]
```diff
@@ -8121,3 +8138,56 @@ func validateNodeSelectorTermHasOnlyAdditions(newTerm, oldTerm core.NodeSelector
...
+	if !validSupplementalGroupsPolicies.Has(*supplementalGroupsPolicy) {
+		allErrors = append(allErrors, field.NotSupported(fldPath, supplementalGroupsPolicy, sets.List(validSupplementalGroupsPolicies)))
+	}
+	return allErrors
+}
+
+func validateContainerStatusUsers(containerStatuses []core.ContainerStatus, fldPath *field.Path, podOS *core.PodOS) field.ErrorList {
+	allErrors := field.ErrorList{}
+	osName := core.Linux
+	if podOS != nil {
```
Comment (thockin) [inline pkg/apis/core/validation/validation.go:8155]: I'm not super familiar with `core.PodOS` - when is it set?  What happens if it is not set?  Can we get here with this field being nil?  I don't see where scheduler considers it, and I don't see any default value defined.

@dchen1107

**Review by everpeace — COMMENTED** [review]
Addressed comments for no need featuregate checks in validation logics. And rebased to the latest master.

**Comment by thockin** [comment]
/approve for API

Needs sig-node review, please

**Comment by mrunalp** [comment]
I will review for sig-node.

File: pkg/apis/core/validation/validation.go (hunk 21 above)
Comment (mrunalp) [reply pkg/apis/core/validation/validation.go:8149]: It looks like this was introduced to detect Windows in admission plugins. https://github.com/kubernetes/enhancements/pull/2803/files
Scheduler interaction was marked as out of scope in that KEP.

Comment (thockin) [reply pkg/apis/core/validation/validation.go:8149]: So this field is truly optional and advisory.  OK.

File: api/openapi-spec/v3/api__v1_openapi.json [hunk 22]
```diff
@@ -1595,6 +1595,14 @@
             "default": {},
             "description": "State holds details about the container's current condition."
           },
+          "user": {
+            "allOf": [
+              {
+                "$ref": "#/components/schemas/io.k8s.api.core.v1.ContainerUser"
+              }
+            ],
+            "description": "User represents user identitiy information of the first process in the container"
```
Comment (mrunalp) [inline api/openapi-spec/v3/api__v1_openapi.json:1604]: typo: identity

File: pkg/apis/core/types.go [hunk 23]
```diff
@@ -2741,6 +2741,31 @@ type ContainerStatus struct {
 	// +optional
 	// +featureGate=RecursiveReadOnlyMounts
 	VolumeMounts []VolumeMountStatus
+	// User represents user identitiy information of the first process in the container
```
Comment (mrunalp) [inline pkg/apis/core/types.go]: typo: identity

**Comment by mrunalp** [comment]
The node bits look fine. Only other nit I have is making the commits additive vs. fixing up earlier commits.
We can add e2es for this once the runtimes support this. Thanks!

File: api/openapi-spec/v3/api__v1_openapi.json (hunk 22 above)
Comment (everpeace) [reply api/openapi-spec/v3/api__v1_openapi.json:1604]: addressed in [5f0d761](https://github.com/kubernetes/kubernetes/pull/117842/commits/5f0d7613520d87745a229f8cc854bbf0e2e18221)

File: pkg/apis/core/types.go (hunk 23 above)
Comment (everpeace) [reply pkg/apis/core/types.go]: addressed in [5f0d761](https://github.com/kubernetes/kubernetes/pull/117842/commits/5f0d7613520d87745a229f8cc854bbf0e2e18221)

**Comment by everpeace** [comment]
> Only other nit I have is making the commits additive vs. fixing up earlier commits.

I think squashing my commits into one would be nice. Is it ok to squash?? 

> We can add e2es for this once the runtimes support this. Thanks!

For container runtimes supporting this feature, we need to merge this PR first as defined in [CRI API | Feature development](https://github.com/kubernetes/kubernetes/blob/6546c8b/staging/src/k8s.io/cri-api/README.md#feature-development). Thus, e2es will be added in another PR after containerd released the newer version which supports this. Am I right?

**Comment by mrunalp** [comment]
> Thus, e2es will be added in another PR after containerd released the newer version which supports this. Am I right?
Yes, that's right.

> I think squashing my commits into one would be nice. Are we ready to squash now??
The PR looks ready to me.

**Comment by everpeace** [comment]
@thockin @mrunalp Thank you very very much for your review and patient nurturing🙇!

**Comment by thockin** [comment]
No, thank you for your patience and perseverance!
